pkg/storage/stores/indexshipper/downloads/table.go

```diff
@@ -347,18 +382,19 @@ func (t *table) EnsureQueryReadiness(ctx context.Context, userIDs []string) erro
 
 // downloadUserIndexes downloads user specific index files concurrently.
 func (t *table) downloadUserIndexes(ctx context.Context, userIDs []string) error {
-	return concurrency.ForEachJob(ctx, len(userIDs), maxDownloadConcurrency, func(ctx context.Context, idx int) error {
-		indexSet, err := t.getOrCreateIndexSet(ctx, userIDs[idx], false)
-		if err != nil {
-			return err
-		}
+	sets, err := t.getOrCreateIndexSets(ctx, userIDs, false)
+	if err != nil {
+		return err
+	}
 
+	return concurrency.ForEachJob(ctx, len(userIDs), maxDownloadConcurrency, func(ctx context.Context, idx int) error {
+		indexSet := sets[idx]
 		return indexSet.AwaitReady(ctx)
 	})
 }
```

Comment:
we bound the number of index sets we download at a time to `maxDownloadConcurrency`. This change would cause us to download all of them simultaneously. If there are too many index sets to download, it would not leave a lot of resources for the query operation.

`indexSet.Init` function initiates the download operation, and `indexSet.AwaitReady` just blocks the function until the `indexSet` gets ready to query after the initial download of the index. I think it would be better to bind the number of index sets we download at a time.